One config only

Anaphory opened the pull request:
Make beastling work with only one config file, and implement a separate tool for merging and interpolating configs.

This is a de-facto change of API, but we *may* be justified in including it for a minor version number.

Please review carefully.

Fixes #139.
Fixes #121, #56, though in a different way than suggested there.

Commit:
Change config interpolation

Activate interpolation, and make sure it occurs at read time, so things
like `basename = %(basename)s_extension` can work. For this to work,
`Interpolation.before_read` must be actually called *before* writing all
data back to the ConfigParser object, so this also overloads two methods
of the ConfigParser, as per http://stackoverflow.com/a/41724876/1274613.

Commit:
Remove duplication of behaviour from ConfigParser

Commit:
Add description of multiconfigs and interpolation

Add a section in the documentation part on configs concerning how to use
multiple configuration files and how interpolation (exists and) can help
with that.

Fixes #121

Commit:
Add tests for interpolation

Commit:
Interpolate defaults and document

Make default section values available in interpolation, and document
how interpolation works in the configuration doc.

Commit:
Add config files for interpolation tests

Commit:
Fix and test interpolation

Global interpolation works with an explicit [DEFAULT] section, I thought
I had seen code that implied it was the stuff outside any section.

In the current implementation, multiline interpolation can work exactly
as expected, so there are some tests guaranteeing that behaviour now.

Commit:
Start separating config magic and core functions

Support only core functionality (reading one and only one config file)
on the cli, and make the actual parsing of one or multiple config
files (as needed by tests) raise exceptions if files are not found.

Permit passing files or strings or lists of files or strings or dicts
to the configparser, so we can use FileType in the argparser.

Given that file existence is now checked by the ArgumentParser, fix
the expected error codes in the CLI tests.

Move interpolation tests to a different file and skip them, so we have
a reminder for implementing the config magic in a standalone tool.

Anaphory commented:
I don't have an explicit opinion yet on where `configalchemy` should live and whether that's a good name for it. For the moment, its version history and documentation is well-established here, but I'm entirely open to putting it on its own or into some other project like `clldutils` (I believe @xrotwang mentioned that as an option).

Commit:
Create configalchemy module

Commit:
Reflect configalchemy in documentation

The new bits of documentation now mention that configalchemy can be
used for it, but beastling cannot.

This is not true yet (beastling still has a "+" in its ArgumentParser)
and the documentation of both tools should be separated and
cross-linked, but at least the documentation reflects the desired and
available API better now.

Commit:
CHANGE BEASTLING API

Remove option to have multiple configfile arguments from the beastling
command line.

This option was never tested and documented on the master branch and
was depracated and the alternative tested and documented with the
previous commit on this branch.

xrotwang commented on docs/config.rst:
```diff
@@ -229,3 +239,148 @@ Each clock section *must* contain the following parameters, i.e. they are mandat
+
+There is a separate tool called ``configalchemy`` included in the
+``beastling`` distribution. ``configalchemy`` allows you to re-use
+analysis definitions you have written my activating *interpolation*
```
Typo: "by" not "my"

xrotwang commented on setup.py:
```diff
@@ -39,3 +39,27 @@
     package_data={'beastling': ['data/*']},
 )
+
+setup(
```
Wouldn't it be simpler to only add another `console_scripts` entry point to the BEASTling setup?

Anaphory replied:
Yes! I just thought this would give a cleaner history until I'm sure it goes into Beastling or somewhere different.

xrotwang commented:
I think having this functionality inside BEASTling is ok for now - I don't have any other use cases which would call for such functionality standalone, or within a library like `clldutils`. I just would reduce it to one console script, and maybe call it `configling` :)

Anaphory commented:
I'll think about it. I can have the file serving as both a read-time interpolation library *and* a command line script, without too much noise for either, so that sounds reasonable.

(`configling` does not convince me – the tool is neither beastling-specific, nor does it have anything to do with linguistics, nor is it a clever or lame pun.)

Commit:
Add new files forgotten in the last commit

Also fix a typo

Commit:
Reduce configalchemy to one file

Commit:
Clean up setup for single module

Commit:
Remove missed debugger call

This call lead to nosetest (which executed the break point,
but did't display it) hanging.

Anaphory commented:
I seriously don't know how I can find the lines in the diff that are missed using the new codecov interface.

xrotwang commented:
What does "nosetests" show you locally? (You may have to downgrade to nose
4.0, above, missing lines aren't reported by default.)

On Mon, Feb 13, 2017 at 10:52 AM, Gereon Kaiping <notifications@github.com>
wrote:

> I seriously don't know how I can find the lines in the diff that are
> missed using the new codecov interface.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lmaurits/BEASTling/pull/140#issuecomment-279340062>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA1HKDVbIPWfZZojV18BYogRRmbQ8zWhks5rcCfggaJpZM4L8WGs>
> .
>

Commit:
Follow PEP8

Anaphory commented:
Found the pesky line that was not covered and the reason it was not covered. Merge at your convenience, does not need to be any time soon – There are API changes in here that want to be considered.

While I did touch some tests, I don't consider this changing tests in the API changing sense I discussed earlier, because the changes concern obvious bugs in tests (wrong filenames, missing method call brackets, etc.) – I therefore think this change would be appropriate for 1.3.0 and would not need to wait for 2.0.0.

But if you want more time to consider options with respect to
 - [ ] The config merging tool's name
 - [ ] Closer or more loose connection between configalchemy and beastling

we can of course wait until paper and 1.3.0 are well out of the way and milestone this for 1.4.0

Commit:
Add a test

codecov.io was rightly complaining that we did not test
beastling/extractor.py:58 – The test that was supposed to hit that
line generated invalid XML, so beastling died before encountering any
problems with embedded config comments.

The old test now explicitly looks for invalid XML, while a new test
with a valid XML file should cover that line.

Commit:
Make read-from-stdin the default if no file provided

As suggested by @xrotwang in #141, "no input argument => read from
stdin" is a nice standard convention for commands reading only one
file, so this adds this feature to one-file-argument beastling.

Anaphory commented:
@lmaurits This is restructuring the code instead of adding features, so it's for 1.5?

lmaurits commented:
I suppose so, although I presume it's a relatively minor restructure (still haven't actually RTFSed this yet!), so maybe we could sneak it in for 1.4 if the other stuff gets taken care of quickly?

Commit:
Make language…groups available for calibration, incl. Glottolog

Calibration clades are now defined in terms of language_groups, but
the language_group method provides a way to access explicitly defined
groups with fallback to glottolog clades (including language families
by name).

Adresses #181 (but does not fix it, by far.)

xrotwang commented on configalchemy.py:
```diff
@@ -0,0 +1,259 @@
+        val: line1
+          line2
+          line1
+          line2
```
Just reading this now. I don't know what the status of this PR is/should be. But I don't find this kind of interpolation really transparent. It means that you can change sets by re-ordering the lines in the config.

lmaurits replied:
I have to admit I was never *super* enthusiastic about this idea.  I really like the way BEASTling currently allows you to "stack" different configuration files, with rightmost files overwriting leftmost ones.  It's quick and easy and intuitive, at least IMHO.

Anaphory replied:
The reason I wanted something like this is to be able to write

    [admin]
    basename = %(basename)_with_grammar
    [model grammar]
    data = grammar.json

and the behaviour given in the patch is the one that allows this behaviour and generalizes it in the way I found obvious or easy to implement. It may not be both, I think it's only the latter.

Anaphory replied:
So, independently of how we handle multi-line entries, @lmaurits this was meant to make the stacking nicer, not less nice.

lmaurits commented:
I need to close this PR in order to make it possible to delete the `develop` branch, which we agreed to do some time ago (see #230).  If we still want to consider this change hopefully it can be rebased onto `master` without too much effort?  Sorry, I know it's a little rude to close this unilaterally, but I want to remove the risk of future work accidentally being based off `develop` ASAP.
